### ISSUE TITLE
Using TRAMP with ivy is cumbersome

### ISSUE THREAD

**Issue by tsdh** [abo-abo/swiper#59]
Using `ivy-mode` you can hardly use TRAMP to find remote files or local files as different user.  The problem is that as soon as you type `/` every input you entered before vanishes.  That makes it impossible to enter a TRAMP file name like `/user@host.domain.tld:/home/user/.emacs` because after typing the `/` after `host.domain.tld:` you end up with just the input `/` again.

When you enter `/user@host.domain.tld:` and then type <kbd>C-j</kbd> it somehow seems to work but your HOME on the machine is opened in dired.  I would have expected that I can further complete the dirs and files on the remote machine.

This is not really pressing, though.  But I think there will always be special cases where ivy has problems so would it be possible to add some `ivy-fallback` command with a key binding which falls back to `completing-read-default` just for the current completion?

**Comment by abo-abo** [comment]
Falling back to `completing-read-default` would be giving up, I think I can still fix this without it.

I was thinking about switching out <kbd>/</kbd> for a while. Now, I want <kbd>//</kbd> to enter root, while a single <kbd>/</kbd> should simply self-insert.  Would this solve your problem?

**Comment by tsdh** [comment]
Well, a bit but not completely.  Now I can type a TRAMP file name and `RET` will successfully open the file.  But there is no completion support whatsoever.  For example, when I do `C-x C-f /ssh:user@host.domain.tld:/ho C-j` it'll find literally the file `/ho` on that machine instead of completing to `/ssh:user@host.domain.tld:/home/`.  However, when I do `C-x C-f` in the buffer of that remote file, the prompt starts out with `/scp:user@host.domain.tld:/` and now the remote candidates (`/etc/`, `/bin/`, ...) are all displayed and completion works fine.

So it kinda works but it's hardly as convenient as with the default completing read function.  There you'd get these kinds of completions:

```
C-x C-f /ssh:user@host.domain.tld:/path/to/file
         1        2               3
```

1. TRAMP method completion (`adb:`, `ssh:`, `scp:`, `smb:`, etc.)
2. Host/domain name completion
3. Remote file completion

So with the default completing read function you can essentially complete your way to the remote files you are frequently editing without having to type too much yourself.

**Comment by abo-abo** [comment]
I'll try to work some more specifically to make tramp work.

**Comment by tsdh** [comment]
As said, no need to hurry.

**Comment by abo-abo** [comment]
> TRAMP method completion (adb:, ssh:, scp:, smb:, etc.)

This I get for both default and ivy.

> Host/domain name completion

This I don't get even for default. Can you clarify how to configure Emacs got get Host/domain name completion?

> Remote file completion

It should be possible to type now: `/ssh:user@domain.com:` <kbd>C-j</kbd> and get completion.

**Comment by tsdh** [comment]
I think host/domain name completion works by checking `/etc/hosts` and `~/.ssh/known_hosts`.  You don't need any special config for that, i.e., I can just do `emacs -Q` and then `C-x C-f /ssh:<TAB>` and get a list of host names and IP addresses.

With the new version, when I do `C-x C-f /ssh:jif<C-j>` the prompt changes to `3 Find File: /ssh:jif` where the path is no input anymore but fontified using `minibuffer-prompt`, and I actually do get completions `fybox:`, `fybuntu:`, and `fybuntu.tsdh.org:`.  The actual valid host names are `jiffy...`, so the completions actually show not the candidates but the remainder of what would make the input a candidate.  When I hit `C-j` with the first "candidate" selected, I get a new buffer with file name `/ssh:jif/fybox:` which is wrong of course.

When I do `C-x C-f /ssh:jiffybox:<C-j>`, I do get the correct remote candidates so that works fine now.

**Comment by abo-abo** [comment]
I'm having trouble configuring this. Can you give me an example of adding such an auto-completed host?  Also, is the user name part of the completion?

**Comment by tsdh** [comment]
Sure.  Ah, the variable `tramp-completion-function-alist` determines for which TRAMP method, e.g., ssh, which files are used for gathering completions.  The value here with `emacs -Q` is

```
(("fcp"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("psftp"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("pscp"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("plink"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("ksu"
  (tramp-parse-passwd "/etc/passwd"))
 ("sudo"
  (tramp-parse-passwd "/etc/passwd"))
 ("su"
  (tramp-parse-passwd "/etc/passwd"))
 ("nc"
  (tramp-parse-hosts "/etc/hosts"))
 ("telnet"
  (tramp-parse-hosts "/etc/hosts"))
 ("sshx"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("ssh"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("rsync"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("scpx"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("scp"
  (tramp-parse-shosts "~/.ssh/known_hosts")
  (tramp-parse-sconfig "~/.ssh/config"))
 ("smb"
  (tramp-parse-netrc "~/.netrc"))
 ("ftp"
  (tramp-parse-netrc "~/.netrc"))
 ("adb"
  (tramp-adb-parse-device-names "")))
```

so for ssh it'll check `~/.ssh/known_hosts`.  So as soon as you have done `ssh localhost` on a terminal, you should be offered the host name completion `localhost`.  And of course you should also get completions for all other `~/.ssh/known_hosts` entries.

**Comment by abo-abo** [comment]
Should work better now. I tried `/ssh:` <kbd>C-j</kbd> and `/sudo:` <kbd>C-j</kbd>: the completion works fine.

**Comment by tsdh** [comment]
It used to work once for me, and now not anymore.  One thing I've noticed that the host name completion only works after `/ssh:` but not after `/ssh:username@`.  And the file completion after `/ssh:host:` didn't work, i.e., TRAMP has successfully logged in but no candidates were shown.  I then hit `C-j` which opened a dired buffer for my remote $HOME.

But that was only some brief testing.  I currently have no time to come up with good recipes for reproduction.

**Comment by abo-abo** [comment]
> But that was only some brief testing. I currently have no time to come up with good recipes for reproduction.

Can you check again if it's working? I'm not having any problems with this.

**Comment by tsdh** [comment]
Yes, sure.

`C-x C-f /ssh:<tab><tab>` indeed now completes host names.  What's a bit confusing is that with the first tab I get the confirm prompt, and that even stays after the second tab which shows the host names as candidates.  IMO, the confirm prompt's meaning is "you are one tab, C-j, or return away from creating the file or buffer whose name will be the input".  That is, usually I'd stop tabbing then and correct my input.  But here, the input is correct and the hosts do exist.

`C-x C-f /ssh:user@<tab><tab>` works like above and now the host candidates have the form `user@host`.  So that works, too.

`C-x C-f /ssh:lin<tab>` given confirm prompt.  Another tab deletes my input and shows the known host names as candidates.  There I select `linux.uni-koblenz.de` and hit tab again.  Now the prompt is `0  Find File: /ssh:linux.uni-koblenz.de:` and no input.  Yet another tab makes TRAMP connect and query for my password.  I enter it, and then dired is opened with my remote HOME.  I had expected to get the files and directories in my HOME on the remote as the next completion candidates.  Well, actually, I should have gotten those even without another tab.

**Comment by tsdh** [comment]
Oh, when I do the last test for the second time, I immediately get remote file/dir completions after the prompt changed to `Find File: /ssh:linux.uni-koblenz.de:`.

Testing a bit more: it seems that I do get remote candidates if I either have logged on previously and the passphrase is still cached or when I connect to a machine where I can login without password but with public key authorization.

So basically there are three small issue left:

1. You get the confirm prompt even though the next tab offers the remotes as completion.
2. The your input will be deleted when host or user@host completions kick in.
3. When logging in with passphrase you don't get completions but the remote HOME is found in dired immediately.

But anyway, it's quite usable now.  Not perfect but workable.

**Comment by abo-abo** [comment]
I think the issue here is the division between <kbd>TAB</kbd> (`ivy-partial-or-done`) and <kbd>C-j</kbd> (`ivy-alt-done`).

The actual TRAMP stuff only happens with <kbd>C-j</kbd>, in this case <kbd>TAB TAB</kbd> is just a shortcut for <kbd>C-j</kbd>.  Could you revise your issues, using <kbd>C-j</kbd> instead of <kbd>TAB</kbd>. Then tell me which functionality (if any) should be moved straight to <kbd>TAB</kbd>, without an additional <kbd>TAB</kbd> needed as now.

**Comment by tsdh** [comment]
I can make a PR if you agree that's TRTTD.

**Comment by abo-abo** [comment]
I don't know that TRTTD is, but I like the PR idea.

**Comment by tsdh** [comment]
The Right Thing To Do. :-)

**Comment by abo-abo** [comment]
Well, I don't really use <kbd>TAB</kbd> (always <kbd>C-j</kbd>), so I'm fine with whatever change to <kbd>TAB</kbd> you suggest.

**Comment by abo-abo** [comment]
> But anyway, it's quite usable now. Not perfect but workable.

Fine, I'll close the issue then, since it has become too verbose to see what is already fixed, and what still needs to be fixed.

> The your input will be deleted when host or user@host completions kick in.

Can you open a new issue with a recipe for this?

> When logging in with passphrase you don't get completions but the remote HOME is found in dired immediately.

I think it's fine if it does this, isn't it? I think `ido` has the same behavior.

**Comment by tsdh** [comment]
Yes and almost yes (ido's completions come after the passphrase of course, but it doesn't simply open some file matching the input immediately but offers the completions after logging in).  I'll open separate issues later.

**Comment by eclig** [comment]
I'm having trouble opening files using tramp's `sudo` protocol: `/sudo::/var/log/messages`. When I type the second `/', the one just after the two colons, the previous input gets removed.  Based on the comments above I was able to find a workaround: `/sudo:localhost:/var/log/messages'.

**Comment by abo-abo** [comment]
@eclig `/sudo::/etc/passwd` works fine for me. `/sudo:://` doesn't, but there's no reason it should.

See if you can reproduce with `emacs -Q`. In that case, provide more version details.

As a suggestion, I always do `/sudo::` <kbd>RET</kbd> - the completion continues after that.

**Comment by eclig** [comment]
Thanks for looking at it, @abo-abo!

I'm using a self-compiled Emacs 24.5 and start it using `emacs -Q`.  The behavior was the same with Emacs 24.4.

After that I go to the directory containing swiper (~/.emacs.d/elpa/swiper-20160418.723) and load `colir.el` and then `ivy.el` (using `dired-do-load`).  I start `ivy-mode` using `M-x ivy-mode`.

After that I just use `C-x C-f` and then enter `//sudo::/`.  The first slash clears the input as expected.  But so does the third one too.  Pressing ENTER after the second colon (ie without typing the third `/`) works, but brings me to root's home directory.

**Comment by eclig** [comment]
Could you reproduce it in the meantime?  Thanks!

**Comment by abo-abo** [comment]
@eclig If you want sudo from the current directory, enter `/sudo::`. Otherwise, enter `///sudo::`. The method that you use, `//sudo::` indeed doesn't work.

**Comment by alyato** [comment]
hi @tsdh Im newer for emacs and want to set the `tramp` by using `tramp-get-completion-function`
The following config is 

> (setq tramp-default-method "ssh")
 (tramp-set-completion-function "ssh"
                                 '((tramp-parse-sconfig "/etc/ssh_config")
                                   (tramp-parse-sconfig "~/.ssh/config")))

when i start the emacs ,it show me 

> Symbol's function definition is void: tramp-set-completion-function

Then i call `c-h f tramp-set-completion-function`, it doesn't show this function.
But i check the link https://www.gnu.org/software/emacs/manual/html_node/tramp/Customizing-Completion.html
![tram](https://user-images.githubusercontent.com/6815071/27783682-c773db6e-6009-11e7-9b13-a061d06d25c3.png)

Do you give me some adivce? Thanks.

**Comment by danrharms** [comment]
Tramp has not been loaded yet. You need to 

(require 'tramp) 

before calling your functions. Or wrap them inside

(with-eval-after-load 'tramp
    ...

**Comment by alyato** [comment]
@danrharms Thanks.It runs well.
